web-gateway/src/main/java/controllers/TransactionController.java

```diff
@@ -77,28 +78,29 @@ public static Call transactionsPage(TransactionInfoStatus status, int page, int
-                    CompletionStage<TransactionInfo> transactionFuture = transactionService.getTransaction(itemId).handleRequestHeader(authenticate(user)).invoke();
-                    return transactionFuture.handle((transaction, exception) -> {
-                        if (exception == null) {
-                            Optional<User> seller = Optional.empty();
-                            Optional<User> winner = Optional.empty();
-                            for (User u : nav.getUsers()) {
-                                if (transaction.getCreator().equals(u.getId())) {
-                                    seller = Optional.of(u);
-                                }
-                                if (transaction.getWinner().equals(u.getId())) {
-                                    winner = Optional.of(u);
+                            UUID itemId = UUID.fromString(id);
+                            CompletionStage<TransactionInfo> transactionFuture = transactionService.getTransaction(itemId).handleRequestHeader(authenticate(user)).invoke();
+
+                           return transactionFuture.handle((transaction, exception) -> {
+                               if (exception == null) {
+
+                                   UUID sellerId = transaction.getCreator();
+                                   UUID winnerId = transaction.getWinner();
+                                  getUser(sellerId).thenCombine(getUser(winnerId),
+                                       (seller, winner) -> {
+
+                                           Currency currency = Currency.valueOf(transaction.getItemData().getCurrencyId());
+                                           return ok(views.html.transaction.render(showInlineInstruction, Optional.of(transaction), user, Optional.of(seller), Optional.of(winner), Optional.of(currency), Optional.empty(), (Nav) nav));
+                                       });
+                           } else {
+                                    String msg = exception.getCause().getMessage();
+                                    return ok(views.html.transaction.render(showInlineInstruction, Optional.empty(), user, Optional.empty(), Optional.empty(), Optional.empty(), Optional.of(msg),(Nav) nav));
                                 }
-                            }
-                            Currency currency = Currency.valueOf(transaction.getItemData().getCurrencyId());
-                            return ok(views.html.transaction.render(showInlineInstruction, Optional.of(transaction), user, seller, winner, Optional.of(currency), Optional.empty(), nav));
-                        } else {
-                            String msg = exception.getCause().getMessage();
-                            return ok(views.html.transaction.render(showInlineInstruction, Optional.empty(), user, Optional.empty(), Optional.empty(), Optional.empty(), Optional.of(msg), nav));
+                                return null;
```

Comment:
Your Happy branch ( `if (exception == null)`)  returns a `CompletionStage<Result>` while the `else` branch returns a `Result`. That won't compile.

The problem here is that there is no combinator in `CompletionStage<T>`  that will accept `Function2<T, Throwable, CompletionStage<U>>` as an argument. That implies that you can't implement the happy path and the recover in the same lambda.
I think a possible solution to this situation is this code:

```scala
                    return transactionFuture.thenCompose(transaction -> {
                            //...
                        CompletionStage<Result> resultCompletionStage =
                                        getUser(sellerId).thenCombine(
                                                 getUser(winnerId),
                                                 (seller, winner) -> {
                                                       Currency currency = //...
                                                       return ok(views.html.transaction.render(// ...
                                                 }
                                        );
                        // the happy path branch returns a CompletionStage<Result> that is unwrapped in the
                        // `thenCompose` above so the returned type of `transactionFuture.thenCompose(...`
                        // is not `CompletionStage<CompletionStage<Result>>` but
                        // `CompletionStage<Result>`
                        return resultCompletionStage;
                    }).exceptionally(exception -> {
                           // this `exceptionally` is slightly different than before, since previous code only captured 
                           //  exceptions on `transactionFuture`, while this captures exceptions in the whole 
                           //  `transactionFuture.thenCompose(...)`. This means that before, if there was an exception 
                           //  `getUser` it was't captured while this code will
      
                            String msg = exception.getCause().getMessage();
                            return ok(views.html.transaction.render(//...
                        }
                    );
```